src/solve/infer/unify.rs

```diff
@@ -166,19 +166,22 @@ impl<'t> Unifier<'t> {
     }
 
     fn unify_forall_tys(&mut self, ty1: &QuantifiedTy, ty2: &QuantifiedTy) -> Fallible<()> {
-        // for<'a...> T == for<'b...> U where 'a != 'b
+        // for<'a...> T == for<'b...> U
         //
         // if:
         //
         // for<'a...> exists<'b...> T == U &&
         // for<'b...> exists<'a...> T == U
+        //
+        // Here we only check for<'a...> exists<'b...> T == U,
+        // can someone smart comment why this is sufficient?
```

Comment:
@nikomatsakis I got curious about this check, I feel like there are assumptions related to regions that are not being spelled out, do you remember what's going on here?

Reply:
Huh, it sure doesn't look right to me. We should try to concoct some test. I imagine that this equality would come out as true:

```
for<'a, 'b> fn(&'a u32, &'b u32) = for<'a> fn(&'a u32, &'a u32)
```

It's actually debatable whether this is correct -- as described in https://github.com/rust-lang/rust/issues/33684, I think we actually **should** consider those two types as equal, which may require changes to the unification algorithm implemented here.

But something like *this* is clearly false, and I imagine that some ordering of it may succeed today?

```
for<'a, 'b> fn(&'a u32) -> &'b u32 = for<'a> fn(&'a u32) -> &'a u32
```

Reply:
I don't think either of those equalities would hold under this algorithm since they require `'a = 'b`, and didn't understand why we'd want the first one to hold.

Reply:
The first one is tricky. I at first thought it should not hold, but I think that this wrong. The starting point is that if `T <: U` and `U <: T` then `T = U`. And "clearly"  the more general form should be a "subtype" of the less general form (since you can trivially instantiate it that way):

```
for<'a, 'b> fn(&'a u32, &'b u32) <: for<'c> fn(&'c u32, &'c u32)
```

But what about the reverse? Here, because `'a` and `'b` only appear in contravariant positions, we can actually consider `for<'c> fn(&'c, &'c)` to be a subtype of `for<'a, 'b> fn(&'a, &'b)` because, no matter what regions `'a` and `'b` are instantated with when the function is called, `'c` will just be the intersection of them. Put another way, consider that functions with those two signatures can freely call one another:

```rust
fn foo<'a>(x: &'a u32, y: &'a u32) {
    bar(x, y) // OK: 'b and 'c can both be bound to `'a` (or, in fact, the call site)
}

fn bar<'b, 'c>(x: &'b u32, y: &'c u32) {
    foo(x, y) // OK: 'a is inferred to the call site
}
```

([This is the same argument I made before here](https://github.com/rust-lang/rust/issues/32330#issuecomment-202536977))

Reply:
Ah thanks for reviving that comment, the unification is too restrictive then, do we even have that differentiation between a late-bound and early-bound universal in chalk? Should I open an issue to track this?

Reply:
OK so there are two separate things, right @leodasvacas? The first is the case where we *want* equality to be true and the second is one where we do not... I still think this algorithm is incomplete as impl'd, but it'd be nice to find examples of both where it fails.

> do we even have that differentiation between a late-bound and early-bound universal in chalk?

We do, yes:

https://github.com/rust-lang-nursery/chalk/blob/62038e04b1cd19a77b3dd2a51903435617c1c333/src/ir.rs#L435

At least if I understood your question.

Reply:
I'm not sure how to call it, but I mean that difference between an universal lifetime in inputs vs in outputs of a function. Can we construct a type in chalk that represents that difference?

Reply:
Ah, I see. Well, the difference there arises from **variance**, and you're right that Chalk just doesn't model that at all, which is part of the problem here I suppose. We may have to start.